### DISCUSSION TITLE
Using Rails 7.1 async_count

### DISCUSSION THREAD

**Discussion by Timmitry** [ddnexus/pagy#561] [Feature Requests]
Using `pagy(collection)` with ActiveRecord will trigger a COUNT-query. With Rails 7.1, a new `.async_count` method was added, which can be used instead of `.count`. I was wondering if this could be used here:

https://github.com/ddnexus/pagy/blob/836fb75b97da6c4a8edf02c3fd04bd658b014c70/lib/pagy/extras/arel.rb#L24-L27

The counting will only be performed asynchronous if the application is configured to use async database queries, otherwise nothing should change.

**Comment by benkoshy** [comment]
Thank you for your suggestion.

I would need to look at the PR before making commenting with intelligence. 

However, at first glance, I can see no objection to it (apart from dealing with backwards compatibility)?

I will have to return to this issue.

**Comment by ddnexus** [comment]
Sorry, I forgot to comment on this one.

The main concern is that pagy tries to be agnostic, so no rails code should be used in the gem code.

The `pagy` method is not completely agnostic to all  ORMs (`count.(:all)` is not), however it is working with mostly all versions of AR, hence it's an useful exception, which is also easy to override.

From what you say it looks like the introduction of the `async_count` would break all the older apps that use the default `pagy` method, so I think that the app should explicitly override it.

In other words, it would be safer as an opt-in overriding, rather than as an opt-out default feature.

**Reply by Timmitry** [reply]
Hey @ddnexus, thanks for your reply. I totally understand your concerns, the Gem should of course work for all supported versions.

How would you feel about checking the used ActiveRecord version in the code, and using `count(:all)` or `async_count(:all)` depending on whether it is >= 7.1? 

...I might be able to provide a PR, but can't promise anything right now...

**Reply by ddnexus** [reply]
A condition to decide which environment the app runs in should be configuration code that runs only once at app startup.

Adding it in that method would run a check at every single request, for every single app, forever.

It wouldn't let me sleep at night. 🤷

**Comment by benkoshy** [comment]
FYI - if you want to override, consider the ideas below:

https://ddnexus.github.io/pagy/docs/how-to/#override-pagy-methods

Feel free to re-open if I've missed the mark, or you have further info to add.
